File: app/src/main/java/com/teixeira/subtitles/activities/BaseActivity.java
package com.teixeira.subtitles.activities;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import com.google.android.material.color.MaterialColors;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.teixeira.subtitles.R;

abstract class BaseActivity extends AppCompatActivity {

  protected abstract View bindView();

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    getWindow().setStatusBarColor(getStatusBarColor());
    getWindow().setNavigationBarColor(getNavigationBarColor());

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
      getWindow().setNavigationBarDividerColor(getNavigationBarDividerColor());
    }
    super.onCreate(savedInstanceState);
    setContentView(bindView());

    requestPermissions();
  }

  protected int getStatusBarColor() {
    return MaterialColors.getColor(this, com.google.android.material.R.attr.colorSurface, 0);
  }

  protected int getNavigationBarColor() {
    return MaterialColors.getColor(this, com.google.android.material.R.attr.colorSurface, 0);
  }

  protected int getNavigationBarDividerColor() {
    return MaterialColors.getColor(this, com.google.android.material.R.attr.colorOutlineVariant, 0);
  }

  private void requestPermissions() {
    if (!isPermissionsGranted()) {
      new MaterialAlertDialogBuilder(this)
          .setTitle(R.string.permission_request)
          .setMessage(R.string.permission_request_message)
          .setPositiveButton(
              R.string.grant,
              (d, w) -> {
                String[] permissions;
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                  permissions = new String[] {Manifest.permission.READ_MEDIA_VIDEO};
                } else {
                  permissions =
                      new String[] {
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE
                      };
                }
                ActivityCompat.requestPermissions(this, permissions, 1000);
              })
          .setNegativeButton(R.string.no, null)
          .show();
    }
  }

  protected boolean isPermissionsGranted() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      return ContextCompat.checkSelfPermission(this, Manifest.permission.READ_MEDIA_VIDEO)
          == PackageManager.PERMISSION_GRANTED;
    } else {
      return ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
          == PackageManager.PERMISSION_GRANTED;
    }
  }
}

File: app/src/main/java/com/teixeira0x/subtypo/ui/activity/BaseActivity.kt
/*
 * This file is part of SubTypo.
 *
 * SubTypo is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SubTypo is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with SubTypo.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira0x.subtypo.ui.activity

import android.os.Build
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.R
import com.google.android.material.color.MaterialColors
import com.teixeira0x.subtypo.App
import com.teixeira0x.subtypo.ui.fragment.dialog.ProgressDialogFragment

/**
 * Base activity for all application activities.
 *
 * @author Felipe Teixeira
 */
abstract class BaseActivity : AppCompatActivity() {

  private var progressDialog: ProgressDialogFragment? = null

  protected open val statusBarColor: Int
    get() = MaterialColors.getColor(this, R.attr.colorSurface, 0)

  protected open val navigationBarColor: Int
    get() = MaterialColors.getColor(this, R.attr.colorSurface, 0)

  protected open val navigationBarDividerColor: Int
    get() = MaterialColors.getColor(this, R.attr.colorOutlineVariant, 0)

  protected val app: App
    get() = App.instance

  protected abstract fun bindView(): View

  override fun onCreate(savedInstanceState: Bundle?) {
    window?.apply {
      this.statusBarColor = this@BaseActivity.statusBarColor
      this.navigationBarColor = this@BaseActivity.navigationBarColor
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
        this.navigationBarDividerColor =
          this@BaseActivity.navigationBarDividerColor
      }
    }
    super.onCreate(savedInstanceState)
    setContentView(bindView())
  }

  protected fun showProgressDialog(message: String? = null) {
    ProgressDialogFragment.newInstance(message)
      .also {
        dismissProgressDialog() // Dismiss previous dialog
        progressDialog = it
      }
      .show(supportFragmentManager, null)
  }

  protected fun dismissProgressDialog() {
    progressDialog?.also { it.dismiss() }
    progressDialog = null
  }
}
